### ISSUE TITLE
date_group_offset in write

### ISSUE THREAD

**Issue by yohplala** [dask/fastparquet#542]
Target of this feature request is extending `write(file_scheme='hive')` so that data _merging_ becomes possible.
Proposal is to let the user specify an `index_offsets` to split a dataset instead of a fixed number of rows through `row_group_offsets`.

```python
fastparquet.write(df, file_scheme='hive', index_offset='1D', append=True)
```

This feature would be allowed only for DataFrame with a `DateTimeIndex`.
`index_offset` could as well be a `pandas.tseries.offsets.DateOffset`

With this approach:
- parquet filename would become `part.`**`index_offset_start`**`.parquet`
- with `append=True`, any new data back in time (older than existing data already in the dataset) would be merged in the corresponding `part.*.parquet` file
- metadata would be updated as required if needed

3 scenarri can be distinguished when `append=True`
- new data is newer than any data stored in existing dataset, and is newer by more than 1 increment over the latest `index_offset`
   It is then simply written in a new `part.*.parquet` file

- new data is newer than any data stored in existing dataset, but is no older by 1 increment than the latest `index_offset`
   Then, the lattest data already stored in existing dataset need to be read so that both this data and the new one are merged, and then the lattest exsting `part.*.parquet` file is overwritten (possibly new files are also written, if newer data extends more than 1 increment versus the latest `index_offset`)

- new data is older than some data already stored in existing dataset
   Then, any existing parquet files with overlapping data need to be read. This data and the new data are merged, and the loaded parquet files are overwritten

I would propose to write a script to mimic this behaviour 1st to clarify what this option does and submit it to you.
If it seems relevant to you, this could be completed with the metadata update most notably.
Do you think this makes sense?

This is related to a discussion I just opened on the Dask github: https://github.com/dask/dask/discussions/7012.1
I am starting to doubt this feature already exists so I am proposing to write a first rough draft.
Any feedback is more than welcome :).
And even if not integrated in fastparquet, I think I would develop it as it appears needed within my project :) (except if I see another way to deal with this matter of updating data stored in parquet files)

Thanks for your help and support!

**Comment by martindurant** [comment]
The simple part of this I completely agree with: providing offsets other than simple integers for making the row-groups. This would be easy to implement. There would be no need to name the files to reflect the offsets, but it also wouldn't hurt, so long as they still order correctly to `glob`.

The rest of the idea about updating is far more complex, as you have described. I wonder how other frameworks handle this. Here are some thoughts.
- Firstly, you must ensure that the existing dataset is already ordered according to the column you want to work with. 
- Next, updating the data in a single row-group. I don't think there's any way to do this without reading the row group, concatenating in memory and writing again (because the columns are stored separately). As an aside, pyarrow specifically allows for loading *multiple* row-groups from one file within a dataset as one task, which would map well to an append operation and reasing with dask
- updating metadata in-place in the `_metadata` file; this is doable, but all new code. You can currently recreate the metadata from all the data files, but have to read the metadata of every file to do it. The metadata object itself can be quite large in memory, and the current thrift implementation does not allow for parsing through to some point and updating/adding values.

btw: I did not get a notification from your discussion at dask/dask, so I'm would think others weren't notified either. You should probably raise an issue to get more attention. The dask situation is complicated by also having pyarrow to think about, which for the time being has *two* working APIs.

**Comment by yohplala** [comment]
Hi @martindurant,
Thanks a lot for your answer.
I would like to propose you a first draft (maybe first as a separate script, then working on integrating it in fastparquet).
You would have a clear view about my intent, but already you have things in mind, yes!

`Firstly, you must ensure that the existing dataset is already ordered according to the column you want to work with.`
Yes, this would be ensured by the fact that the user use `index_offset` parameter. From there, a `sort_index` is necessarily carried out before the write within the writer class.

`Next, updating the data in a single row-group. I don't think there's any way to do this without reading the row group, concatenating in memory and writing again (because the columns are stored separately).`
I confirm that the implementation I am working on load the existing row group (stored in a single file) one by one to update them before writing the updated row group. It proceeds then to the next row group, on and on till the input DataFrame to be written gets exhausted.

`updating metadata in-place in the _metadata file; this is doable, but all new code. You can currently recreate the metadata from all the data files, but have to read the metadata of every file to do it. The metadata object itself can be quite large in memory, and the current thrift implementation does not allow for parsing through to some point and updating/adding values.` 
Ok, I would propose a 1st milestone by re-using what is already available (full rewrite of the _metadata). If later on I see a need for updating this part (me or another user), let's issue a separate ticket for that.

From your answer, my 1st target is to propose you a draft.
From then on, if fastparquet team is ok on this feature, we can move on.

`btw: I did not get a notification from your discussion at dask/dask, so I'm would think others weren't notified either. You should probably raise an issue to get more attention. The dask situation is complicated by also having pyarrow to think about, which for the time being has two working APIs.`
I am not sure if this topic should be better raised in dask or in fastparquet.
I have started to delve into it with fastparquet and pandas libs. I will implement a dataframe processing by chunk, so it does be in the spirit of dask (or so is my understanding), but dask appears to me like a big project to contribute to. Maybe discussions are easier within smaller scale projet like fastparquet, which does aim to interact with parquet files (or so is my understanding) :)

Thanks again!

**Comment by martindurant** [comment]
> I am not sure if this topic should be better raised in dask or in fastparquet.

I imagine Dask would be interested, if it can be made to work. The difference, of course, is that dask writes happen in parallel - which might be fine, if the update data is already sorted the same way as the target dataset. It would be fine to implement here and later to expose to Dask.

**Comment by yohplala** [comment]
Hi @martindurant 
Please, find [here](https://github.com/yohplala/fastparquet_append) a draft implementation.
I made the README as a small set of examples / tutorial to showcase the `date_offset` feature.
Please, I would like to ask you to take some time to read `merge` function docstring as it is where I have listed all features, limitations, and todos.

To sum up the todos (short term):
- to proceed to a proper integration into fastparquet, and what I have not implemented by lack of knowledge about fastparquet, is storing the `date_offset`information in the `_common_metadata`. This offset value has to stay constant (it is imposed, not like `row_group_offsets` that can be a list). The user has thus to define it at the 1st write, but then, it should not be required for any new append. Please, could you advise me on the appropriate way to do so? With this advice, I would then start to prepare a PR.
- please, also, do not hesitate to share any advices on the code,
   - where you would see the different functions (in which source files?) or if you would me to split the code in different functions.
   - if I can use fastparquet internal functions instead of some I have created (naming the parquet files, `os` functions to create directory, handling metadata...)

In the longer term, there are probably limitations you would like me to work on. While not necessary for my own project, I will understand your requirements and I am ready to go further. I am thinking most notably to:
- accepting a `date_offset` not only on index, but possibly on a column instead, that the user will specify. This should be fine to implement. It is a bit more code /a bit more complex, but clearly not unfeasible.

Then, in the longer longer term, there are limitations I do not intend in the short term to spend time on, but you should be aware of them.
- proposed partitioning is based on `pandas.DateOffset`. There could probably be other 'measures' that could be used for partitioning. I don't think it is relevant to work on this until someone states a need for that, and clarify a use case.
- `overwrite` mode. At the moment, I am complexifying the `append` mode (`append=True`). Within `append` mode, only new data is added, or existing data can also be modified. Data cannot be removed. This could be made feasible within the frame of an `overwrite` mode (possibly replacing `append`parameter by a `write_mode` parameter with values `None`, `append`, `overwrite`). At the moment, it is not a priority for me.
- metadata update. We discussed this already. At this moment, they are fully re-built. I would be interested in trying to update only the part that needs to be updated, but I have a big lack of knowledge about this, and till I don't see a clear limitation because of that, I may not look into it in the short term.

On my side, I have still a few tests to do (mentionned in `tests.py`). I will confirm correct working of partitinoning with bigger `pandas.DateOffset`: 1D, 1MS most notably.
Then, if you are fine with the feature as it is poposed in this draft, I will try to see if I can propose a PR.

I will welcome any feedback, please, do not hesitate.
Have a good evening, bests

**Comment by yohplala** [comment]
Hi @martindurant 
Happy new year to you and the whole fastparquet team! :)

Martin, I checked the code with following test cases:
- DataFrame with DatetimeIndex
- `date_offset` : `2H`, `1D`, `1W`, `1M` : I solved 2 bugs along this process :)

I further wanted to validate the code using DataFrame with PeriodIndex but I am currently blocked by a bug that I think is external to my code. I raised the ticket #543 in fastparquet github for this blocking bug.
Please, can you have a look?

Thanks for your help!
Bests,

PS: I have updated the lists of tests I made in https://github.com/yohplala/fastparquet_append/blob/main/tests.py
PPS: only for chatting, just for you to know, I am that guy to whom you answered in [this SO question](https://stackoverflow.com/questions/65308299/repartition-by-size-with-dask-not-producing-files-with-expected-size) :)

**Comment by martindurant** [comment]
Thanks for the work, I should have a chance to look on Monday. Also, thanks for including so much doc and comments; how hard do you think it would be to merge what you've done into this repo, tests and all?

**Comment by yohplala** [comment]
> Thanks for the work, I should have a chance to look on Monday. Also, thanks for including so much doc and comments;

Thanks @martindurant. Ah ah, no worries, it is my usual way of working. I am here 'negotiating' to integrate a new feature ;).

> how hard do you think it would be to merge what you've done into this repo, tests and all?

I am absolutely unable to say, I have implemented this as I would have done, would I implement it in my own project, that is to say a feature external to fastparquet.
To integrate it properly to fastparquet, I am missing an adequate knowledge about fastparquet.
Tomorrow, I had in mind to workout the ability for the user to specify a column for date grouping (instead of forcing the user to have it in the index).
On Monday, if you can have a look, I would like to propose you the following:
- either you see that the integration would not be much work and you can try
- or you see it is some work, I can try doing it under your guidance. If so, please, I would be glad if you could pinpoint what you would change in the code to have it properly integrated?
    - most notably, there are 2 information we need (IMHO) to record in the metadata. Can you show me maybe somewhere in the code how recording info in metadata is done? These 2 information are:
         - as already said: the `date_offset` (that I think actually to rename to `date_group_offset` similar to `row_group_offsets`, but without s because it does not accept a list)
         - not said, but required if we can let the user specify the column containing the dates to be used, a `group_on` parameter (for the user to specify which column to use - what I intend to have a look to tomorrow)

With this information, I can take a look at how integrating this.
To have it integrated, I would probably need a week to have the code more or less clean. Then the tests... hmmm I am no professional software engineer, and I have no knowledge about the standard library you are probably using for this. I would have to learn this as well (I could no worries - in my own 'private' project, I do have non-regression tests, but it is a simple script I have developped on my own).

Sorry to write so much.
Depending on your first review on Monday, yes you can count on me to help further for a proper integration. :)

**Comment by yohplala** [comment]
Hi @martindurant, I hope you are fine.
Ok, I updated the draft implementation of the 'append' strategy based on 'date_group_offset' parameter so that it now accepts the 'group_on' parameter. With this parameter, the user can name a column to be used for date grouping instead of using the DataFrame index (which remains the default if 'group_on' is not provided)

One comment with this option. After appending operations, when reloading the full Dataframe, the index gets ...meaningless... ? For example:
```python
       humidity  pressure location           timestamp
index                                                 
0           0.5   90000.0    Tokyo 2019-12-01 01:59:00
1           0.3  100000.0    Paris 2020-01-02 01:58:00
0           0.4  190000.0    Paris 2020-01-02 01:59:00
2           0.9   95000.0    Milan 2020-02-03 02:59:00
3           0.8  200000.0    Paris 2020-02-03 03:59:00
4           1.1   95000.0    Paris 2020-03-03 02:59:00
```

I am not sure it is anyway of much use, but then, I prefer to raise the question. Is there any guideline on how the default index should look like in such a case, with 'file_scheme='hive''?
Does it restart to 0 at each row group?
Should it pursue index of the previous file? (which would defeat the point of the current append strategy which aims to make possible to append locally new data without to rewrite a DataFrame in its whole completness)
In this latter case, i would suggest to let it as it is, but then, tell me what you think would be best.

Have a good evening, bests,

**Comment by yohplala** [comment]
Between, I just tried to use a PeriodIndex as a column in fastparquet, but the write function returns the same error message as when it is an index (I wanted to validate the append strategy using a PeriodIndex as column thanks to the 'group_on' parameter)

**Comment by martindurant** [comment]
>  Is there any guideline on how the default index should look like in such a case, with 'file_scheme='hive''?

No there isn't, the index is, effectively, meaningless. However, we *do* know how many rows are in each row-group, so it would make sense to create a RangeIndex for each part rather than make a fresh default index starting from zero. Again, this would be some work, and would have some implications for users of fastparquet that want to filter out row-groups (e.g., dask might do this), when the value of the index might become ambiguous.

The list of rows per row-group could be extracted as
```
[rg.num_rows for rg in pf.row_groups]
```
(where `pf` is the `ParquetFile` instance)
and the indexes might be
```
from itertools import accumulate, chain
[pd.RangeIndex(start=i, stop=j) for i, j in zip(
    chain((0, ), accumulate(rg.num_rows for rg in pf.row_groups)), 
    accumulate(rg.num_rows for rg in pf.row_groups)
)]
```

**Comment by yohplala** [comment]
Hi @martindurant 

I have stated to have a closer look to fastparquet and I start to see that I will need some time.
I have a few questions, please can you help me clarify?
- 1st, the metadata update. I have raised this question already. I can see where are handled metadata per row group, but not it there is a way to store global metadata. Can you share any pointer?

- 2nd, I am starting to wonder if it might not be wiser/simpler to benefit from 'partition_on` approach. The periods could be managed through folders as it is already possible (for instance considering 3 layers: year, month, day of month)
Now the need / new feature of _updating_ existing data remains. If re-using 'partition_on` approach, no index is provided by the user to sort the Dataframe content. It means that the update is made by concatenating the new data to the whole content the folder having existing data overlapping with the new one.

As a workaround of the difficulty that represents concatenating one Dataframe to several row groups, I am thinking that a `sort_on` parameter could be made compulsory. This way, we could achieve sorting.

Do you think this might be a direction worth studying?
It may ease the implementation.
I can actually try to implement this 2nd version based on `partition_on` approach. I will keep you informed.
Please, do not hesitate to share your mind about it.
Bests,

**Comment by martindurant** [comment]
> I can see where are handled metadata per row group, but not if there is a way to store global metadata.

They go into the `_metadata` file. We have cod to gather the pieces from many files and create this, but to update it in-place without having to touch all the files is possible - but this would need to be implemented.

> if it might not be wiser/simpler to benefit from partition_on approach

If you can partition, for example, by month or quarter, then this is fine, and I think already works (at least via dask); but you get new files for all new data, and so end up with many files after repeated appends. Also, this doesn't solve the original problem of arbitrary time splits or those that cannot be treated as categorical - I don't think it's flexible enough for you.

**Comment by yohplala** [comment]
Hi @martindurant, just to let you know, I have started ported my draft implementation into [`writer.py`](https://github.com/dask/fastparquet/blob/master/fastparquet/writer.py).
Once re-implemented, I think I will need your help for a thorough review :)
Bests

**Comment by yohplala** [comment]
Hi @martindurant,
Please, I would need an advice. Is there a method to read a part file 'directly'?

At the moment, to read the content of a parquet part file as a DataFrame, I am using:
```python
                        existing = pf.to_pandas(
                            filters=[(date_index_name, '>=', start),
                                     (date_index_name, '<=', end)])
```

Thing is I directly know the name of the part file.
- I have been thinking to use `ParquetFile.read_row_group_file(rg, columns, categories)` but this method does not accept a filename, and I don't know what to use for `rg`.
- This method calls `core.read_row_group_file(fn, rg, columns, categories)` which seems more fitting, but here again, I don't understand what is `rg` and this method does not return anything actually. (I don't understand the way `df` is set)

Please, would you have any advice or code snippet to retrieve the content of a part parquet file?

Thanks for your help.
Bests,

**Comment by yohplala** [comment]
Ok, I think I got it, noticing the `assign` parameter.

Meanwhile, I have yet another question. Please, my understanding is that successive row groups have the same columns / column name. This is part of the schema, right? One column cannot be omitted/renamed/added with file scheme `hive`, right?

I am asking because to read a row group from a parquet file, once you have the filename, with current code, you need to forward statistics of corresponding row group as well. It means you have 1st to loop over the row group list to find the corresponding one.

Once you have it, it then ends up with the function `[read_row_group_arrays](https://github.com/dask/fastparquet/blob/b37766d34ac0b4374874a60dd94f937ab9c27ef5/fastparquet/core.py#L308)` in `core.py`.

But the parameter `rg` appears (if my understanding is correct) to be forwarded as input only to retrive column names. Please, can you confirm?

This is the motivation of my question: if the schema enforces the same columns (number and names) for all row groups, then I can simply retrieve column names from the 1st row group, right?
If this understanding is right, then I can come up with a 'simpler' function that will cover my need, no need to made a loop over the row groups to find which one is corresponding to my parquet part file (which would be what I am forced to do with current code).

Please, can you confirm?
Thanks for your help. Bests

**Comment by martindurant** [comment]
OK, so (again) it's a little complicated.

Each data file contains in its thrift-encoded metadata footer a copy of the schema, row-group(s) information and any additional metadata such as the pandas stuff. Typically, there is exactly one row-group. Therefore, you can read this file independently of the rest of the dataset by opening it as you would a whole directory - but this will cause the footer to be parsed. Each file indeed contains the same schema and metadata, but different row-group information. Actually, "schema evolution" is allowed in theory (e.g., adding columns or upcasting types) but this is rare and would break for your append idea.

The _metadata file contains the same schema and metadata as the data files, and a copy of all the row-group information blocks for all files, in a single list (in the same thrift-encoded block as usual) and no data. So fastparquet would normally read this file, and because the row groups say which file their data is contained within, and give byte offsets, there is no need to read the footers of the files. This can be a huge time saver. This is why, if you were to append data to any one of the data files, and so change the bytes offsets, you would need to update both the footer of that file and the _metadata file too.

In fastparquet, you can view all these objects that were decoded from thrift:
```
pf = ParquetFile(file-or-directory)
pf.fmd # everything
pf.row_groups # list of row group objects; the statistics, file names and offsets live in the columns of each element
pf.key_value_metadata # parquet metadata as a dict
pf.statistics # dict of stats values, computed on demand
pf.schema # inferred hierarchical schema, containing the original objects in `.schema_elements`
```

So for example, working with the test dataset "split":
```
pf = fastparquet.ParquetFile("split")
pf2 = fastparquet.ParquetFile("split/" + pf.row_groups[0].columns[0].file_path)
assert pf.schema == pf2.schema
assert pf.row_groups[0].columns[0].file_offset == pf2.row_groups[0].columns[0].file_offset
assert pf2.row_groups[0].columns[0].file_path is None # the data is in the same file as this footer
```

**Comment by yohplala** [comment]
Hi @martindurant 
Thanks for taking the time to present all this information.
I am sorry, I have had little spare time to make progress until yesterday evening.
With the work done yesterday evening, I think I am now done something like 75% into the integration of the version 1 for this feature.

Metadata stuff is tackled. Actually, it appears to me that no big work is needed because I am following / slightly updating what is already in `write` function (metadata update starting at [line 974](https://github.com/dask/fastparquet/blob/d738d4256db2c25624889417e72af6a6d59f30f8/fastparquet/writer.py#L974) to 978).
With your explanation, I am more confident in the updates I am making, thanks a lot! From first tests, it appears to be ok (just solved a bug this morning :)).

So what remains to be completed for this version 1 is:
- direct reading of parquet part files when concatenating to existing data,
- and of course thorough testing, re-using the test plan I have already built for the draft version.

In this 1st version, what I am planning NOT to introduce is the following:
- adding new data to metadata:
    - date offset
    - column name where is the date to be used for updating,
    User will have to remind this data to make an update with new data.
- compatibility with `file_scheme=simple`:
   I started to think about it, and the conclusion I have come up with is that I don't see the point of using this `update` feature when file is simple. In this case, were we to update a row group, then the full file would get rewritten anyway. So for me, useless to propose it.
- compatibility with `partition_on`:
  There does be interest to have this compatibility though. So later on, once a 1st version is working, I think I will come back to this. 

Bests

**Comment by yohplala** [comment]
Hi @martindurant 
I am having a look at creating a python test file.
Up to now, to test my work on a separate script, I was re-installing fastparquet after each modification with the command line `python3 setup.py  install`.

But now, in my test script, I need to import `tempdir` thanks to this import:
`from fastparquet.test.util import tempdir`

However, I get this error:
`ModuleNotFoundError: No module named 'fastparquet.test'`

I am suspecting that the install mode I am using exclude the test module.
I can see in `setup.py`following line:
`    exclude_package_data={'fastparquet': ['test/*']},`
So this might indeed be the cause.

Please, what is the command line to be used to install fastparquet in 'develop' mode?
(meanwhile your answer, I guess I will simply comment out this line and see if it works)

Thanks for the advice, bests,

**Comment by martindurant** [comment]
- adding new data to metadata:
  - date offset
  - column name where is the date to be used for updating,

I don't think there's any reason to store this in the metadata, since we can easily see whether the column of interest, as specified by the user at update time, is indeed sorted as required. We error if not. It would normally be the index, right?

- compatibility with file_scheme=simple:

agreed

- compatibility with partition_on:

agreed

**Comment by yohplala** [comment]
Hi @martindurant 
__It would normally be the index, right?__
Yes, this is tested with [`test_write_with_rgp_by_date_as_index`](https://github.com/yohplala/fastparquet/blob/46f3982a5c722de2bf707e72416661370d9fbaa8/fastparquet/test/test_rgp_by_date.py#L25)

__I don't think there's any reason to store this in the metadata, since we can easily see whether the column of interest, as specified by the user at update time, is indeed sorted as required.__
I don't think we understand each other on this.
In the code I have written, you may notice that I make checks on the data, and raise exception if data does not fit.
I may have a tendency of over checking input data, which is maybe not so much fastparquet philosophy compatible (don't know if there is any philosophy on this though... :)).

I would store this data for data check.
Once `date_offset` and `date_col` have been defined at the 1st write, they have not to be changed in next appendings. It will not work otherwise. If the user wants to change `date_offset`, then it has to rewrite the whole dataset to operate this change.
So, by storing this data in metadata, then:
- either we make this data not needed as input for further append
- or if it is provided by the user, and does not match the data in the metadata, we issue an exception
This is the idea behind this bullet point.

**Comment by yohplala** [comment]
Closing as partially replaced by 'overwrite' feature.
